text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+           `node-proxy-agent` is an npm package.
+    4. AWS CDK v2 will automatically pick up new minor versions of the above
+       libraries.
+
+After the above is implemented, this mechanism can be extended to include
+additional Lambda Layers with different dependencies. The solution should be
+implemented in such a way that new dependencies can be added with the below
+steps.
+1. Add a new directory in the `aws-cdk` monorepo: `@aws-cdk/lambda-layer-MyNewDependency`.
+2. Implement the MyNewDependencyLayer API.
+3. If the dependency is from npm, then updates will happen automatically. If it
+   is from some other source, then extend the automation
+   from the above steps to also update the new dependency.
+
+Once this solution is implemented, it will be possible to write more concrete
+steps. And those will be included in the Contributing Guide.
+
+##### What about customers executing their CDK apps in network-restricted environments?
+
+This question raises an important drawback of this solution. It will result in a
+breaking change for customers who are using these Lambda Layers in
+network-restricted environments. If customers are able to use the CDK CLI in
+their network-restricted environments, then it is reasonable to assume they were
+able to acquire the `aws-cdk` package from npm somehow. This implies customers
+will be able to also acquire the `@aws-cdk/lambda-layer-X` packages from npm. They can either
+make sure the necessary packages are already installed and available in the
+`node_modules` directory where their CDK applications are executed, or include
+them in a private npm registry.
+
+Unfortunately, there is no way to remove these large files and host them
+somewhere else without causing a breaking change for these customers. We will
+publish a CLI notice directly to potentially-affected customers of this change
+well in advance of releasing it. And, advertise the change in advance of release
+on the aws-cdk GitHub repository.
+
+##### What about customers in ADC regions?
+
+Today, there is a mechanism for publishing AWS CDK into ADC (Amazon Dedicated
+Cloud) regions. This mechanism includes bundling the direct dependencies of the
+CDK and publishing them into the ADC regions, e.g. `constructs`. We will add the
+`@aws-cdk/lambda-layer-X` packages to this process, and when customers use the aws-cdk npm
+package in ADC regions, the right dependencies will already be present in the
+`node_modules` directory.
+
+#### .jsii Files
+
+There are two files, .jsii.tabl.json and .jsii, bundled in `aws-cdk-lib` for the
+JSII runtime to work for any non-NodeJS languages. And, to enable code and
+documentation generation from the published npm package. Today, they make up 42%
+of the package size. And in the future, after adding more Lambda Layers, the .jsii files would
+compose 30% of the package size.
+
+We will compress the .jsii.tabl.json and .jsii files. Compressing with gzip
```

Comment:
Is this valuable? I would argue it doesn't matter, since the download size will always be the already-compressed size. 

If we're going to do anything, we should probably try to find a more efficient encoding of the assembly format (with less redundancy), potentially stop pretty-printing it.

Reply:
The "unpacked" npm size is also important. I think this is a concrete step we should take right now to improve that. We can still pursue a more efficient assembly format as well. I am not sure what the options are there, and haven't done any investigation into it. Do you have any ideas?

Reply:
Even though download size is the more important metric, reducing unpacked size will have some benefits for disk usage of projects (and the aforementioned 250MB Lambda footprint limit--even though it's possible to work around that limit, let's try to not force people to have to do that). Compressing is an easy way to get that size down, so why not.

One thing I missed originally is that the JSII manifest embedded in NPM packages will not (typically) be used at runtime. It will only be loaded by jsii build tooling--and this is especially true for tablets: so gzipping them will not usually have a big performance impact.

The only place we need to pay attention is jsii client languages. The Python bundle currently looks like this (containing the full NPM package, this pattern is repeated for all jsii languages):

```
┌───────────────────────────────────────────────────┐                    
│                  aws_cdk_lib.whl                  │                    
├───────────────────────────────────────────────────┤                    
│                                                   │                    
│                                                   │                    
│                                                   │                    
│               ┌───────────────────────────────┐   │                    
│               │     aws-cdk-lib.jsii.tgz      │   │                    
│               ├───────────────────────────────┤   │                    
│               │                               │   │                    
│   ┌──────┐    │  ┌──────┐   ┌─────────────┐   │   │       jsii kernel  
│   │ *.py │    │  │ *.js │   │    .jsii    │◀──┼───┼─────   needs this  
│   └──────┘    │  └──────┘   └─────────────┘   │   │           file     
│               │                               │   │                    
│               │   ┌───────────────────────┐   │   │                    
│               │   │    .jsii.tabl.json    │◀──┼───┼─────  nobody needs 
│               │   └───────────────────────┘   │   │        this file.  
│               │                               │   │         kill it.   
│               └───────────────────────────────┘   │                    
│                                                   │                    
└───────────────────────────────────────────────────┘                    
```

Two observations:

- The `.jsii.tabl.json` file is in there, which nobody needs. We can totally remove it to save a bunch of space.
- The jsii kernel is going to need to load the `.jsii` file, which turns into a `.jsii.gz` file if we go through with the compression strategy, paying a small performance penalty to load and gunzip the library on startup (we need to quantify how much this costs!). Either that or we put the uncompressed version in the jsii client archive, but then our disk space savings only materialize for TypeScript users, not any other language users.

Reply:
The tablet file is used by ConstructHub to not time out when generating documentation...

Reply:
Thanks for confirming that we need to keep the table file, @RomainMuller .